### ISSUE TITLE
Support for grouping targets

### ISSUE THREAD

**Issue by SirLynix** [xmake-io/xmake#1026]
### Is your feature request related to a problem? Please describe.
I'm working on Visual Studio 2019, until now I was using [premake](https://github.com/premake/premake-core) to generate my projects. I'm considering replacing it by xmake (because your tool is truly awesome, congrats!) and I'm trying to replicate one feature premake had, groups.

Grouping targets (or projects in premake terms) corresponds to folders in Visual Studio:

### Describe the solution you'd like
Some API like `target.set_group` to specify a group.

### Describe alternatives you've considered
Using a prefix in the target name.

### Additional context
Screenshots:
With folders (using Premake):
![image](https://user-images.githubusercontent.com/3002461/98677457-53cff380-235d-11eb-9520-743837ef200c.png)

Without folders (using xmake):
![image](https://user-images.githubusercontent.com/3002461/98677526-677b5a00-235d-11eb-99c0-b43ab9a3daf6.png)

**Comment by waruqi** [comment]
Are you using `xmake project -k vs` plugin or `xmake project -k vsxmake` plugin?

**Comment by SirLynix** [comment]
Here I used the `vsxmake` plugin, I know `vs` is deprecated (although it would be nice to add folders to it too).

**Comment by waruqi** [comment]
We can consider adding a new configuration in future versions: `set_values("vsxmake.target_group", "xxx")` to set the group name of each target. 

```lua
target("xx")
    set_values("vsxmake.target_group", "xxx")
```

But for version 2.3.9, we will not support this feature. For version 2.3.9, we focus on improving package management.

**Comment by SirLynix** [comment]
No problem.
Wouldn't it be better to support groups as as target parameter? Groups could be implements in multiple projects, as far as I know Visual Studio and XCode both support them. But adding support to makefiles could be done as well.

For example if I have two projects in the same group (let's say `"common"`), makefiles could support `make common`

And of course, if a project doesn't support groups they can be safely ignored.

**Comment by waruqi** [comment]
set_groups("xxx", "yyy")?   xmake build xxx yyy? 

It seems a good idea, I will consider it, but it may not be supported soon. This involves a lot of changes, you may need to wait some time.

**Comment by SirLynix** [comment]
Why not target.set_group?

Sure, no problem. I'm already glad you like the idea. 😄

**Comment by waruqi** [comment]
> Why not target.set_group?
> 
> Sure, no problem. I'm already glad you like the idea. 😄

only for target.

```lua
target()
    set_group()
```

**Comment by SirLynix** [comment]
Yes that's what I was suggesting. Seems great! 👍

**Comment by waruqi** [comment]
After the current 2.3.9 version is released, I will support it in the next version 2.5.1 and there won't be 2.4.x

**Comment by AntoineJT** [comment]
Okay, I've opened a similar issue, so I report my ideas here. (You can see the original text here: https://github.com/xmake-io/xmake/issues/1172)

I think the way Premake handles this is not the best way to do.
I think it can be better handled by doing as target does:
```lua
group("name")
    target("first_target")
        ...
    target("second_target")
        ...
group_end()
```

Alternatively, it may be handled like this:
```lua
target("first_target")
    ...
target("second_target")
    ...
target_end()

group("name", "first_target", "second_target")
#or
group("name", {"first_target", "second_target"})
```
or maybe it can be handled [the same way as Premake does](https://github.com/premake/premake-core/wiki/group), but I think the better solution here for users would be the first I propose.

**Comment by waruqi** [comment]
> group("name")
    target("first_target")
        ...
    target("second_target")
        ...
group_end()

The grouping feature is currently only used in the project generator plugin, and the focus of xmake is on `xmake build`, so I don’t consider adding an additional `group()` scope api to implement it for now. I will use the `set_groups("xxx", "yyy")` interface to specify target grouping.

**Comment by SirLynix** [comment]
Maybe groups could be more useful for xmake, like to build a set of targets. Maybe also set a after build file (like something to execute only after all targets in a group are built). Wouldn't that be interesting?

**Comment by waruqi** [comment]
> Maybe groups could be more useful for xmake, like to build a set of targets. Maybe also set a after build file (like something to execute only after all targets in a group are built). Wouldn't that be interesting?

This can be achieved using `add_rules` for the same purpose.

**Comment by AntoineJT** [comment]
At least, can we consider
```lua
create_group("name", "first_target", "second_target")
# or
create_group("name", {"first_target", "second_target"})
```
to avoid duplication everywhere?
Maybe it can be added with `target.set_group` too, as an alternative way to group targets

**Comment by waruqi** [comment]
I will consider to add `target.set_group`

**Comment by waruqi** [comment]
I have supported for vs project on group branch, but I still don’t support vsxmake, you can try vs project first. https://github.com/xmake-io/xmake/pull/1174

```lua
add_rules("mode.debug", "mode.release")

target("test1")
    set_kind("binary")
    add_files("src/*.cpp")
    set_group("group1")

target("test2")
    set_kind("binary")
    add_files("src/*.cpp")
    set_group("group1")

target("test3")
    set_kind("binary")
    add_files("src/*.cpp")
    set_group("group1/group2")

target("test4")
    set_kind("binary")
    add_files("src/*.cpp")
    set_group("group3/group4")

target("test5")
    set_kind("binary")
    add_files("src/*.cpp")

target("test6")
    set_kind("binary")
    add_files("src/*.cpp")
```

![Snip20201228_7](https://user-images.githubusercontent.com/151335/103217563-d2a3de80-4953-11eb-95b5-b41d16a1abfa.png)

**Comment by AntoineJT** [comment]
Yeah, it works, that's great.

![image](https://user-images.githubusercontent.com/26577763/103218487-69a07480-491b-11eb-874b-fb61528f885a.png)
![image](https://user-images.githubusercontent.com/26577763/103218689-09f69900-491c-11eb-948a-011f4f6431a6.png)

Don't worry about the missing SDL2.dll, that's normal I guess with the vs2019 solution. I just need to add it manually in the build folder.

I'm using vsxmake for my C++ learning project so I will comment out the set_group things until you add it to vsxmake too.

**Comment by waruqi** [comment]
I have also supported it for vsxmake project and merged into dev branch.
